File: question/neetcode150/binary_search/search-a-2d-matrix.py
from typing import List

class Solution:
    def searchMatrix(self, a: List[List[int]], t: int) -> bool:
        n,m = len(a), len(a[0])

        i = 0
        l = n*m

        while l:
            while i+l < n*m and a[(i+l)//m][(i+l)%m] <= t:
                i+=l
            l//=2

        return a[i//m][i%m] == t

File: question/neetcode150/binary_search/search-a-2d-matrix.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& a, int t) {
        int n = a.size(), m = a[0].size();

        int k = 0;

        for(int i=n*m;i>0;i/=2){
            while((k+i)<n*m and a[(k+i)/m][(k+i)%m]<=t){
                k+=i;
            }
        }

        return a[k/m][k%m] == t;
    }
};